paddle/fluid/imperative/hooks.h

```diff
@@ -0,0 +1,231 @@
+ * VarBase hooks
+ * - PyGradAccumulatorPostHook (Implement later): used for user-defined leaf
+ * python VarBase hooks
+ */
+
+/** [ Hook Package classes ]
+ *
+ * @note  [Why need hook package classes?]
+ *
+ *        There are 2 purposes for adding Hook Package here:
+ *
+ *        1. Make the code implementation cleaner.
+ *
+ *          If there are no Hook Package, we need to add 3 hook vector into
+ *          VariableWrapper, 1 hook vector into OpBase, 2 hook vector into
+ *          GradientAccumulator, like:
+ *
+ *          - VariableWrapper:
+ *            std::vector<std::shared_ptr<OpBasePreHook>>
+ *              interior_var_hooks_;
+ *            std::vector<std::shared_ptr<GradAccumulatorPostHook>>
+ *              leaf_var_hooks_;
+ *            std::vector<std::shared_ptr<GradAccumulatorPostHook>>
+ *              backward_hooks_;
+ *
+ *          - OpBase:
+ *            std::vector<std::weak_ptr<OpBasePreHook>>
+ *              interior_var_hooks_;
+ *
+ *          - GradientAccumulator:
+ *            std::vector<std::weak_ptr<GradAccumulatorPostHook>>
+ *              leaf_var_hooks_;
+ *            std::vector<std::weak_ptr<GradAccumulatorPostHook>>
+ *              backward_hooks_;
+ *
+ *          This seems more complicated, and std::vector<std::weak_ptr<...>>
+ *          is not easy to destruct.
+ *
+ *        2. Make the code easier to understand.
+ *
+ *          From these two packages, we can clearly understand that we
+ *          have two types of Hooks, respectively for the interior
+ *          gradient var and leaf gradient var inside the backward
+ *          calculation graph.
+ */
+
+class InteriorVarHookPackage {
+ public:
+  InteriorVarHookPackage() = default;
+
+  void add_hook(std::unique_ptr<OpBasePreHook>&& hook) {
+    hooks_.emplace_back(std::move(hook));
+  }
+
+  const std::vector<std::unique_ptr<OpBasePreHook>>& hooks() const {
+    return hooks_;
+  }
+
+  std::vector<std::unique_ptr<OpBasePreHook>>& hooks() { return hooks_; }
+
+ private:
+  std::vector<std::unique_ptr<OpBasePreHook>> hooks_;
+
+  DISABLE_COPY_AND_ASSIGN(InteriorVarHookPackage);
+};
+
+class LeafVarHookPackage {
+ public:
+  LeafVarHookPackage() = default;
+
+  void add_hook(std::unique_ptr<GradAccumulatorPostHook>&& hook) {
+    hooks_.emplace_back(std::move(hook));
+  }
+
+  const std::vector<std::unique_ptr<GradAccumulatorPostHook>>& hooks() const {
+    return hooks_;
+  }
+
+  std::vector<std::unique_ptr<GradAccumulatorPostHook>>& hooks() {
+    return hooks_;
+  }
+
+  void add_backward_hook(std::unique_ptr<GradAccumulatorPostHook>&& hook) {
+    backward_hooks_.emplace_back(std::move(hook));
+  }
+
+  const std::vector<std::unique_ptr<GradAccumulatorPostHook>>& backward_hooks()
+      const {
+    return backward_hooks_;
+  }
+
+  std::vector<std::unique_ptr<GradAccumulatorPostHook>>& backward_hooks() {
+    return backward_hooks_;
+  }
+
+ private:
+  std::vector<std::unique_ptr<GradAccumulatorPostHook>> hooks_;
```

Comment:
May be it can call 'leaf_var_hooks_' , and can be better distinguished from 'backward_hooks_' . After all, both of them are hooks in backward. Isn't 'backward_hooks_' here for Allreduce/Reduce only?

Reply:
1. my opinion: the class name `LeafVarHookPackage` alreaady hold the`leaf var` info, the `hooks` in `LeafVarHookPackage` are `leaf_var_hooks_ `, using long member name cause information redundancy and also make the interface name longer, such as `LeafVarHookPackage.add_leaf_var_hook()`
2. `backward_hooks_` mean the hooks of `whole backward process, because it relay on leaf var, so we can only put it here now, may be we should add `AccumulateGrad dummy OpNode` and move `backward_hooks_` outside, I wiil perfect the comments here

Reply:
And `backward_hooks_ ` may not only used for Allreduce/Reduce, we should keep scalability here

Reply:
Ok